### PR TITLE
[MAINT] Simplify tests connectome

### PR DESCRIPTION
**Pull request by Remi-Gau**
Relates to https://github.com/nilearn/nilearn/issues/3570

Changes proposed in this pull request:

- refactor tests of the regions module to have fewer assertions in each test

### CODE REVIEW

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 1]
```diff
@@ -475,82 +499,170 @@ def test_connectivity_measure_outputs():
...
+    signals, emp_covs, ledoit_covs = signals_and_covariances
+    kind = "tangent"
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    if isinstance(cov_estimator, LedoitWolf):
+        covs = ledoit_covs
+    elif isinstance(cov_estimator, EmpiricalCovariance):
+        covs = emp_covs
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:553]: See comment above

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 2]
```diff
@@ -475,82 +499,170 @@ def test_connectivity_measure_outputs():
...
+    signals, emp_covs, ledoit_covs = signals_and_covariances
+    kind = "precision"
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    if isinstance(cov_estimator, LedoitWolf):
+        covs = ledoit_covs
+    elif isinstance(cov_estimator, EmpiricalCovariance):
+        covs = emp_covs
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:587]: See comment above

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 3]
```diff
@@ -475,82 +499,170 @@ def test_connectivity_measure_outputs():
...
+    signals, emp_covs, ledoit_covs = signals_and_covariances
+    kind = "correlation"
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    if isinstance(cov_estimator, LedoitWolf):
+        covs = ledoit_covs
+    elif isinstance(cov_estimator, EmpiricalCovariance):
+        covs = emp_covs
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:609]: See comment above

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 4]
```diff
@@ -475,82 +499,170 @@ def test_connectivity_measure_outputs():
...
+    signals, emp_covs, ledoit_covs = signals_and_covariances
+    kind = "partial correlation"
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    if isinstance(cov_estimator, LedoitWolf):
+        covs = ledoit_covs
+    elif isinstance(cov_estimator, EmpiricalCovariance):
+        covs = emp_covs
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:634]: See comment above

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 5]
```diff
@@ -660,15 +814,13 @@ def test_connectivity_measure_outputs():
...
-    n_features = 49
-
     # Generate signals and compute covariances and apply confounds while
     # computing covariances
     signals = []
     for k in range(n_subjects):
         n_samples = 200 + k
         signal, _, confounds = generate_signals(
-            n_features=n_features,
+            n_features=49,
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:823]: Use `N_FEATURES`

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 6]
```diff
@@ -692,6 +844,22 @@ def test_confounds_connectome_measure():
...
+
+def test_confounds_connectome_measure_errors():
+    n_subjects = N_SUBJECTS
+    # Generate signals and compute covariances and apply confounds while
+    # computing covariances
+    signals = []
+    for k in range(n_subjects):
+        n_samples = 200 + k
+        signal, _, confounds = generate_signals(
+            n_features=49,
```
Comment (ymzayek) [inline nilearn/connectome/tests/test_connectivity_matrices.py:856]: Use `N_FEATURES`

**Review by ymzayek — COMMENTED** [review]
Thx @Remi-Gau ! A few minor comments

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py]: done

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 7]
```diff
@@ -453,104 +518,140 @@ def test_connectivity_measure_errors():
...
+    cov_estimator, signals_and_covariances
+):
+    signals, covs = signals_and_covariances
+    kind = "partial correlation"
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    for k, cov_new in enumerate(connectivities):
+        prec = linalg.inv(covs[k])
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:628]: I think we could also parametrize these tests on "kind". It seems like only the assertion part is changing, the rest is the same. So maybe something like this (let me know what you think, it's only a suggestion, I haven't tested whether this works or will end up more readable....):

```python
@pytest.mark.parametrize("cov_estimator", [EmpiricalCovariance(), LedoitWolf()])
@pytest.mark.parametrize("kind", ["covariance", "correlation", "tangent", "precision", "partial correlation"])
def test_connectivity_measure(cov_estimator, kind, signals_and_covariance):
    signals, covs = signals_and_covariances
    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
    connectivities = conn_measure.fit_transform(signals)
    connections = (connectivities, conn_measure, covs)
    _assert_connectivities(connections, kind)

def _assert_connectivities(connections: tuple, kind: str) -> bool:
    if kind == "tangent":
        return _assert_tangent_connectivities(connections)
    if kind == "correlation":
        return _assert_correlation_connectivities(connections)
    ...

def _assert_tangent_connectivities(connections: tuple) -> bool:
    """Check positive definiteness if expected."""
    connectivities, covs, conn_measure = connections
    for k, cov_new in enumerate(connectivities):
        assert_array_almost_equal(cov_new, cov_new.T)
        gmean_sqrt = _map_eigenvalues(np.sqrt, conn_measure.mean_)
        assert is_spd(gmean_sqrt, decimal=7)
        assert is_spd(conn_measure.whitening_, decimal=7)
        assert_array_almost_equal(conn_measure.whitening_.dot(gmean_sqrt), np.eye(N_FEATURES))
        assert_array_almost_equal(
             gmean_sqrt.dot(_map_eigenvalues(np.exp, cov_new)).dot(gmean_sqrt), covs[k],
         )
    return True

.... 
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 8]
```diff
@@ -692,6 +821,22 @@ def test_confounds_connectome_measure():
     confounds_df = DataFrame(confounds[:10])
     correlation_measure.fit_transform(signals, confounds=confounds_df)
 
+
+def test_confounds_connectome_measure_errors():
+    n_subjects = N_SUBJECTS
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:826]:
```suggestion
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 9]
```diff
@@ -692,6 +821,22 @@ def test_confounds_connectome_measure():
...
+    # Generate signals and compute covariances and apply confounds while
+    # computing covariances
+    signals = []
+    for k in range(n_subjects):
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:830]:
```suggestion
    for k in range(N_SUBJECTS):
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @Remi-Gau !
This is looking good ! 👍 
I made a first pass with few suggestions. Feel free to take or leave.

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 7 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py]: done

**Review by ymzayek — APPROVED** [review]
LGTM @Remi-Gau

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 10]
```diff
@@ -425,208 +587,196 @@ def test_prec_to_partial():
...
+    connectivities = conn_measure.fit_transform(signals)
+
+    if kind == "tangent":
+        _assert_connectivity_tangent(connectivities, conn_measure, covs)
+    elif kind == "precision":
+        _assert_connectivity_precision(connectivities, covs)
+    elif kind == "correlation":
+        _assert_connectivity_correlation(connectivities, cov_estimator, covs)
+    elif kind == "partial correlation":
+        _assert_connectivity_partial_correlation(connectivities, covs)
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:672]: It could be nice to put the definitions of these functions just after this test instead of at the beginning of the file. This way, the users won't have to scroll up and down. WDYT ?

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 11]
```diff
@@ -357,26 +510,31 @@ def test_sym_matrix_to_vec():
...
     sym = vec_to_sym_matrix(vec, diagonal=diagonal)
+
     assert_array_almost_equal(
         sym_matrix_to_vec(sym, discard_diagonal=True), vec
     )
 
     # multiple matrices case when diagonal is included
     vecs = np.asarray([vec, 2.0 * vec, 0.5 * vec])
     syms = vec_to_sym_matrix(vecs)
+
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:531]: I might be missing something, but I feel like we could split these in multiple tests (or a single test with parametrization maybe...). WDYT ?

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 12]
```diff
@@ -306,40 +442,57 @@ def test_geometric_mean_properties():
...
 
 def test_geometric_mean_errors():
     n_features = 5
 
     # Non square input matrix
     mat1 = np.ones((n_features, n_features + 1))
-    pytest.raises(ValueError, _geometric_mean, [mat1])
+
+    with pytest.raises(ValueError, match="Expected a square matrix"):
+        _geometric_mean([mat1])
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:480]: I would do three different tests here, one for non square matrices, one for non spd matrices, and one for different shape matrices.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 13]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+    # Test on sqrt map
+    spd_sqrt = np.array(
+        [[2.0, -1.0, 0.0], [-1.0, 2.0, -1.0], [0.0, -1.0, 2.0]]
+    )
+    spd = spd_sqrt.dot(spd_sqrt)
+    assert_array_almost_equal(_map_eigenvalues(np.sqrt, spd), spd_sqrt)
+
+    # Test on log map
+    spd = np.array([[1.25, 0.75], [0.75, 1.25]])
+    spd_log = np.array([[0.0, log(2.0)], [log(2.0), 0.0]])
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:320]: Same kind of comment, this can be split in three

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 14]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+            same_variance=False,
+        )
+        signals.append(signal)
+        signal -= signal.mean(axis=0)
+    return signals, confounds
+
+
+@pytest.fixture
+def signals():
+    signals, _ = _signals(N_SUBJECTS)
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:258]:
```suggestion
    return _signals(N_SUBJECTS)[0]
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 15]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+    return signals
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:259]:
```suggestion
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 16]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+        assert is_spd(cov_new, decimal=7)
+
+        d = np.sqrt(np.diag(np.diag(covs[k])))
+
+        if cov_estimator == EmpiricalCovariance():
+            assert_array_almost_equal(d.dot(cov_new).dot(d), covs[k])
+        assert_array_almost_equal(np.diag(cov_new), np.ones(N_FEATURES))
+
+
+def _assert_connectivity_partial_correlation(connectivities, covs):
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:227]: This might be some work, but it could be nice to add a docstring for each of these `_assert_connectivity...` functions explaining how they check the connectivity. It's not super trivial (to me at least...) when looking at the linalg code below.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 17]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+    for k, cov_new in enumerate(connectivities):
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:228]: We don't need the index : 

```suggestion
    for cov, cov_new in zip(covs, connectivities):
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 18]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+        prec = linalg.inv(covs[k])
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:229]:
```suggestion
        prec = linalg.inv(cov)
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 19]
```diff
@@ -23,6 +23,18 @@
 from sklearn.covariance import EmpiricalCovariance, LedoitWolf
 from sklearn.utils import check_random_state
 
+CONNECTIVITY_KINDS = [
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:26]: I'd expect a tuple for this (unless there is a reason I didn't see for using a list):

```suggestion
CONNECTIVITY_KINDS = (
```

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 20]
```diff
@@ -23,6 +23,18 @@
...
+    "covariance",
+    "correlation",
+    "tangent",
+    "precision",
+    "partial correlation",
+]
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:32]:
```suggestion
)
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @Remi-Gau !
This is looking good ! 
I made some further suggestions as I think some tests can be broken in different pieces.

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 10 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py:672]: yeah I went with everything at the top rule, but if a helper function is used by only one test it would help indeed.

still don't have a good rule for organizing this.

will try to make sure things are closer to where they are used if used only once

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 11 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py:531]: yes probably.

will slit further

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 16 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py:227]: not trivial to me either: but now that things are split it becomes easier to start deciphering each chunk

will try to add some explanations

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 21]
```diff
@@ -259,6 +187,192 @@ def random_non_singular(p, sing_min=1.0, sing_max=2.0, random_state=0):
...
+
+def test_check_spd():
+    non_sym = np.array([[0, 1], [0, 0]])
+
+    with pytest.raises(
+        ValueError, match="Expected a symmetric positive definite matrix."
+    ):
+        _check_spd(non_sym)
+
+    non_spd = np.ones((3, 3))
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py:295]: nope except that I did not see the opportunity: good catch

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 19 above)
Comment (NicolasGensollen) [reply nilearn/connectome/tests/test_connectivity_matrices.py:26]: That is what I understand also, hence relying on a tuple rather than a list.

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 22]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+    ):
+        """Check output value properties for tangent connectivity measure \
+        that they have the expected relationship \
+        to the input covariance matrices.
+
+        - the geometric mean of the eigenvalues
+          of the mean covariance matrix is positive-definite
+        - the whitening matrix (used to transform the data \
+          also produces a positive-definite matrix
+        """
```
Comment (Remi-Gau) [inline nilearn/connectome/tests/test_connectivity_matrices.py:622]: Adding some plain english explanation here but I would not mind having someone cross check that I want I am adding is correct: my linear algebra is not what it used to be.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 23]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+            ).dot(gmean_sqrt),
+            true_covariance_matrix,
+        )
+
+
+def _assert_connectivity_precision(connectivities, covs):
+    """Estimated precision matrix: \
+    - is positive definite, \
+    - its product with the true covariance matrix \
+      is close to the identity matrix."""
```
Comment (Remi-Gau) [inline nilearn/connectome/tests/test_connectivity_matrices.py:648]: same here

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 24]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+        - values on its diagonal should almost 1
+
+    If the covariance estimator is EmpiricalCovariance,
+    the product of:
+        - the square root of the diagonal of the true covariance matrix,
+        - the estimated covariance matrix,
+        - the square root of the diagonal of the true covariance matrix,
+
+    should be close to the true covariance matrix.
+    """
```
Comment (Remi-Gau) [inline nilearn/connectome/tests/test_connectivity_matrices.py:671]: same here

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 25]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+
+        # expected value
+        partial_corrlelation_matrix = -precision_matrix + 2 * np.diag(
+            np.diag(precision_matrix)
+        )
+
+        assert_array_almost_equal(
+            normalized_partial_correlation_matrix,
+            partial_corrlelation_matrix,
+        )
```
Comment (Remi-Gau) [inline nilearn/connectome/tests/test_connectivity_matrices.py:715]: and here

**Review by Remi-Gau — COMMENTED** [review]
@bthirion @NicolasGensollen 
I have added some comments in plain english and changed some varibles names in places to make things a bit more transparent.
But I would not mind have another pair of eyes to make sure it is correct.

Also let me know if you prefer one big doc string explaining the whole function or more inline comments: right now there is a bit of both.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 26]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+        assert_array_equal(input_covs[k], covs[k])
+
+        assert is_spd(covs[k], decimal=7)
+
+
+def _assert_connectivity_tangent(connectivities, conn_measure, covs):
+    for true_covariance_matrix, estimated_covariance_matrix in zip(
+        covs, connectivities
+    ):
+        """Check output value properties for tangent connectivity measure \
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:614]: It feels a bit weird to have the docstring within the for loop. Either move it after the function declaration, or define another function with this docstring that you will call iteratively.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 27]
```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
...
+        assert_array_almost_equal(
+            estimated_covariance_matrix.dot(true_covariance_matrix),
+            np.eye(N_FEATURES),
+        )
+
+
+def _assert_connectivity_correlation(connectivities, cov_estimator, covs):
+    """Verify that the estimated covariance matrix: \
+        - is symmetric and positive definite
+        - values on its diagonal should almost 1
```
Comment (NicolasGensollen) [inline nilearn/connectome/tests/test_connectivity_matrices.py:662]:
```suggestion
        - has values close to 1 on its diagonal
```

**Review by NicolasGensollen — APPROVED** [review]
Thanks @Remi-Gau !
I'll let @bthirion validate the correctness of the explanations.
In my opinion, the tests are MUCH more readable now, thanks a lot ! 👍

File: nilearn/connectome/tests/test_connectivity_matrices.py (hunk 22 above)
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py:622]: Note that some of the steps in this functions are not explained in the doc string: mostly because I am not even entirely sure what the test actually does.

File: nilearn/connectome/tests/test_connectivity_matrices.py [hunk 28]
```diff
@@ -295,8 +313,72 @@ def test_geometric_mean_properties():
...
+        vals_gmean, vecs_gmean = linalg.eigh(gmean)
+        gmean_inv_sqrt = _form_symmetric(np.sqrt, 1.0 / vals_gmean, vecs_gmean)
+        whitened_mats = [
+            gmean_inv_sqrt.dot(mat).dot(gmean_inv_sqrt) for mat in mats
+        ]
+        logs = [_map_eigenvalues(np.log, w_mat) for w_mat in whitened_mats]
+        logs_mean = np.mean(logs, axis=0)  # Covariant derivative is
+        # - gmean.dot(logms_mean)
+        norm = np.linalg.norm(logs_mean)  # Norm of the covariant derivative on
+        # the tangent space at point gmean
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/connectome/tests/test_connectivity_matrices.py]: good point
moving them

**Comment by Remi-Gau** [comment]
It seems this is good to go. Will merge in 24 hours in case anyone wants to have another look before this goes in.
